### PR TITLE
Merge emerging-optimizers change from dev to main

### PR DESCRIPTION
**Pull request by skyw**
# What does this PR do ?
  - Emerging optimizer registry: New emerging_optimizers.py with registry pattern for Muon, AdaptiveMuon, SOAP, Lion, and future
  optimizers. Muon code consolidated from standalone muon.py (now a backward-compat shim) into the registry framework.
  - Layer-wise optimizer improvements: Simplified LayerWiseDistributedOptimizer init (removed async_allgather param, uses config
  directly), improved error messages for pre-wrapped optimizers.
  - Config updates: New fields (use_layer_wise_distributed_optimizer, SOAP/AdaptiveMuon params), muon_use_nesterov renamed to
  muon_nesterov, AdamOptimizerConfig/SGDOptimizerConfig simplified to aliases.
  - CLI args: Expanded --optimizer choices (soap, adaptive_muon), --muon-nesterov rename, dist_muon deprecation with auto-redirect to
  muon + layer-wise.
  - ChainedOptimizer fix: _split_state_dict now uses model chunk identity mapping instead of sequential offset, fixing state dict
  splitting for mixed optimizer types.

related:
  - #3927 — Add --muon-coefficient-type argument for Muon optimizer
  - #3907 — Add more emerging optimizers
  - #3897 — Fix emerging optimizer init_group for ckpt loading
  - #3813 — Add Lion optimizer support
  - #3737 — Continue emerging optimizer refactoring
  - #3715 — Scaling for MuP over Muon optimizer
  - #3618 — Refactor to support emerging optimizers beyond muon
  - #3524 — Add --overlap-param-gather support for layer-wise optimizer

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

**Comment by skyw** [comment]
@FDecaYed plz take a look.  Pay attention to the muon-scalar-optimizer logic which could be breaking.

And some uv settings may be out of sync. CI returned strange missing package errors.

File: megatron/core/distributed/param_and_grad_buffer.py [hunk 1]
```diff
@@ -243,6 +243,9 @@ def __init__(
         # or bucket.grad_data.
         self.cached_param_buffer_shard_list = [None] * len(self.buckets)
         self.cached_grad_buffer_shard_list = [None] * len(self.buckets)
+        # Track grad mode used to create cached param views. Rebuild if mode changes to avoid
+        # mixing no_grad-created views with in-place updates in grad-enabled mode.
+        self._cached_param_buffer_shards_grad_enabled = None
```
Comment (FDecaYed) [inline megatron/core/distributed/param_and_grad_buffer.py:248]: What is this for? seems this is not used anywhere?

File: megatron/core/optimizer/optimizer.py [hunk 2]
```diff
@@ -1161,20 +1161,26 @@ def _split_state_dict(self, state_dict):
         state_dicts = [None] * len(self.chained_optimizers)
         if state_dict is not None:
             if len(self.model_chunks) == 1:
-                state_dicts[0] = state_dict
+                # When there is only one global model chunk, all sub-optimizers
```
Comment (FDecaYed) [inline megatron/core/optimizer/optimizer.py:1164]: changes from https://github.com/NVIDIA/Megatron-LM/pull/3676
@kunlunl also agree to merge here in one go

File: megatron/core/safe_globals.py [hunk 3]
```diff
@@ -33,6 +33,7 @@
     RerunState,
     BytesIO,
     Signals,
+    torch._C.Generator,
```
Comment (FDecaYed) [inline megatron/core/safe_globals.py:36]: why is this needed?

File: megatron/training/arguments.py [hunk 4]
```diff
@@ -1472,17 +1471,26 @@ def validate_args(args, defaults={}):
             '--no-load-optim with --skip-train --perform-rl-step skips the optimizer; ' \
             '--rl-offload-optimizer-during-inference is incompatible (no optimizer to offload).'
 
-    # Muon optimizer check
-    if 'muon' in args.optimizer:
+    # emerging optimizer check
+    if not hasattr(args, 'use_layer_wise_distributed_optimizer'):
```
Comment (FDecaYed) [inline megatron/training/arguments.py:1475]: maybe just add this as a usual parameter below, and set to default false there? it might help allow downstream properly set it(bridge for example). because sometime they just create args without running code logics here.
not a hard request though

**Review by FDecaYed — APPROVED** [review]
some minor issues. otherwise looks good to me

File: megatron/core/distributed/param_and_grad_buffer.py (hunk 1 above)
Comment (skyw) [reply megatron/core/distributed/param_and_grad_buffer.py:248]: It is picked up from one of the PR. let me check if I picked it wrong.

File: megatron/core/safe_globals.py (hunk 3 above)
Comment (skyw) [reply megatron/core/safe_globals.py:36]: After torch.load default change, this is necessary for loading conventional torch format checkpoint of muon and other emerging optimizers.

Comment (skyw) [reply megatron/core/safe_globals.py:36]: added comment

File: megatron/core/distributed/param_and_grad_buffer.py (hunk 1 above)
Comment (skyw) [reply megatron/core/distributed/param_and_grad_buffer.py:248]: Removed. It was from #3890 and unrelated to optimizer.

**Comment by skyw** [comment]
Closing.
In favor of small PRs instead of combined ones.
